### ISSUE TITLE
BOSH DNS doesn't work on  windows 2016

### ISSUE THREAD

**Issue by aminjam** [cloudfoundry/dns-release#8]
When we deploy the [windows2016 version](https://github.com/cloudfoundry/cf-deployment/blob/master/operations/experimental/use-bosh-dns-for-windows2016-containers.yml) of [bosh dns for containers opsfile](https://github.com/cloudfoundry/cf-deployment/blob/master/operations/experimental/use-bosh-dns-for-containers.yml), it fails to establish outbound connections. The opsfile deploys fine, but unable to stage an app when trying to download the buildpack.

```
Downloading hwc_buildpack...
Downloaded hwc_buildpack (2.4M)
Creating container
Downloading app package...
Successfully created container
Downloaded app package (786.4K)
       HWC version 5.0.0
-----> Installing hwc 5.0.0
-----> Hwc Buildpack version 2.3.9
-----> Installing HWC
       Download [https://buildpacks.cloudfoundry.org/dependencies/hwc/hwc-5.0.0-windows-amd64-89274c25.zip]
       **ERROR** Unable to install HWC: Get https://buildpacks.cloudfoundry.org/dependencies/hwc/hwc-5.0.0-windows-amd64-89274c25.zip: dial tcp: lookup buildpacks.cloudfoundry.org: no such host
main.main()
goroutine 1 [running]:
        /tmp/buildpack-packager141830456/src/hwc/compile/cli/main.go:40 +0x4be
panic: Get https://buildpacks.cloudfoundry.org/dependencies/hwc/hwc-5.0.0-windows-amd64-89274c25.zip: dial tcp: lookup buildpacks.cloudfoundry.org: no such host
Failed to compile droplet: Failed to compile droplet: exit status 2
Exit status 223
Stopping instance 5e39cf25-93d2-4564-8571-25f0e8f53412
Destroying container
Successfully destroyed container

FAILED
Error restarting application: BuildpackCompileFailed
```

We are able to see the the `dns_servers` property applied in the container when running `Get-DNSClientServerAddress`. When we don't use the `use-bosh-dns-for-windows2016-containers.yml` opsfile, dns server list is pulled in from the host and this problem doesn't exist.

**Comment by sunjayBhatia** [comment]
This is not just an issue inside containers, @jvshahid and the Diego team deployed BOSH DNS in their CF with `consul` disabled and the `rep_windows` job was unable to resolve other services.

**Comment by dpb587-pivotal** [comment]
@aminjam would you be able to check whether the buildpacks.cloudfoundry.org query hit the dns server at all by reviewing the bosh-dns logs? That should confirm whether the request left the container and hit the correct DNS server (and if the correct recursor is being used). If it doesn't show up, perhaps the container was not correctly passing traffic? Or if it does show up, would be nice to know if bosh-dns was able to recurse (and to where).

@sunjayBhatia could the non-container case have been a separate issue caused by [#152523777](https://www.pivotaltracker.com/n/projects/1479998/stories/152523777)?

**Comment by jvshahid** [comment]
@dpb587-pivotal i had a conversation with @luan about this bug. i had some notes [here](https://gist.github.com/jvshahid/09304d2ef50629e6531836de014a9254) in particular doing a name lookup doesn't hit the bosh dns agent at all. Worth noting that this was done on the cell not inside a container. Switching the bosh dns agent to use `127.0.0.1` as the listen address instead of `169.254.0.2` fixed the name resolution.

**Comment by sesmith177** [comment]
**BOSH DNS on the Host** 
We believe we have found the problem on Windows 2016. It looks to be a change in how Windows does DNS lookups, described here: https://docs.microsoft.com/en-us/windows-server/networking/dns/what-s-new-in-dns-client

Now, when a DNS lookup is done, the client will first bind to the interface that has that address configured as a DNS server. For example:

1. Ethernet interface has DNS server 169.254.0.2 
2. Loopback interface has IP address 169.254.0.2 
3. Process makes DNS lookup request
4. Windows DNS client service sees DNS server 169.254.0.2 is configured for interface Ethernet
5. Windows DNS client service binds to Ethernet interface and makes the request
6. The UDP packets are never routed to the DNS server and the request times out

In 2012R2, the DNS client would not bind to the Ethernet interface in the above example, allowing the UDP packets to be routed properly

@dpb587-pivotal @jvshahid 

**BOSH DNS in Containers**
The problem here appears to be that the loopback interface on the host cannot be accessed from the containers (they have their own separate interfaces in separate network compartments).

We have been able to give container access to BOSH DNS running on the host by having it listen on a non-loopback interface, namely the NAT interface on the host (created to route traffic to containers). Note that this configuration runs into the same problem with BOSH DNS not working on the host as described above
